### DISCUSSION TITLE
Contracts Verification in Hedera

### DISCUSSION THREAD

**Discussion by damianlluch** [hashgraph/hedera-smart-contracts#27] [Q&A] [answered]
Hello, I have a question. 
Is there any contract verification in Hedera?

I have managed to deploy a contract using the FileAppendTransaction() function.

But in the block explorer I see something that is unintelligible for me...

https://testnet.dragonglass.me/transactions/00460501151656320700698270653

Thank you!

**Comment by Nana-EC** [comment, marked as answer]
Hi @damianlluch,
Thanks for reaching out. I converted your issue to a discussion since it doesn't highlight an issue with this repose code.

By contract verification I presume you mean verifying the entity is present post creation.
In your case all you've done is created a file, not yet a contract.

I'd recommend viewing the [create-a-smart-contract documentation](https://docs.hedera.com/guides/docs/sdks/smart-contracts/create-a-smart-contract).
Specifically you could utilize the `ContractCreateFlow` which does the work of creating a File with the byte code contents and then creating an actual contract that reference these bytes.
To verify you can view the contractID output in any mirror node explorer.

Hope that helps

**Reply by damianlluch** [reply]
Hi @Nana-EC thanks

Ok, so the steps are
1- Creation of the file in Hedera.
2- Deploy the contract from the file.

This is the first time I see this uploading a file, what are the advantages?

By contract verification I mean that the contract is displayed in the block explorer.

**Reply by Nana-EC** [reply]
Correct.
In your initial steps you seem to have already been uploading the file (FileCreate & FileAppend). 
You just hadn't issued a Contract create transaction based off the file byte contents.
The noted code does that for you.
One advantage is working around the transaction size limitation, it also allows for great flexibility of file and contract metadata management.

This is the standard way to create smart contract entities on Hedera.
Let me know if this clarifies things and helps unblock you.

**Reply by damianlluch** [reply]
Thank you.
I was able to create the contract with your help.
0.0.46822981

Now, to verify the contract and see the code in the block explorer what is the best way? I see that I have disabled the option to validate contract in DragonGlass.

**Reply by damianlluch** [reply]
I have one more question for you, as I have to solve the issue of proxies.
In the architecture of my contracts (they are deployed in 4 blockchains), I have a transparent proxy in front of a factory, and then a beacon proxy with a token contract.

With the Truffle and Hardhat libraries, it's simple because I don't care how it does it, it just does it.
Now here, how it works at the bytecode level, I have doubts.
Truffle and Hardhat generate a bytecode in the json that has the ABI. But it has a different format than the one generated by solc-js and I don't understand why.

But I can think of using Truffle or Hardhat to generate the bytecodes, and then deploy them with the SDK, how do you see it?


Format generated by solc-js:

60806040523480156200........

Format generated by Hardhat - Truffle

0x60806040526040516200107..

I'll have to remove the '0x' from it? do you have any explanation?

thanks

**Reply by Nana-EC** [reply]
Great, glad to hear that unblocked you.
You can verify the contract existence on multiple [community mirror node explorers](https://docs.hedera.com/guides/mirrornet/community-mirror-nodes)
I do see your contract on Hashscan - https://hashscan.io/#/testnet/contract/0.0.46822981

Also after circling back it's also present on DragonGlass though - https://testnet.dragonglass.me/hedera/search?q=0.0.46822981
You likely didn't change the network from mainnet to testnet when searching.

**Reply by Nana-EC** [reply]
Hi @damianlluch I'm unclear of your 2nd question, not sure if you're asking a solc-js vs truffle/hardhat question or how to utilize the outputs in the SDK. 
Based on my understanding though you're trying to determine how to use the output in the SDK.
You should just be able to remove the `0x` from the bytecode as you noted and use that in your contract deployment using the SDK. Asides that the hex format of the bytes should be identical.

I believe your original question is answered.
Feel free to start a new discussion question item so we can properly capture your follow up if I didn't also address it here.

**Reply by damianlluch** [reply]
> Hi @damianlluch I'm unclear of your 2nd question, not sure if you're asking a solc-js vs truffle/hardhat question or how to utilize the outputs in the SDK.
Based on my understanding though you're trying to determine how to use the output in the SDK.
You should just be able to remove the 0x from the bytecode as you noted and use that in your contract deployment using the SDK. Asides that the hex format of the bytes should be identical.
> Creo que su pregunta original ha sido respondida.Siéntase libre de comenzar un nuevo elemento de pregunta de discusión para que podamos capturar adecuadamente su seguimiento si no lo abordé también aquí.

Hi @Nana-EC 

I will try to explain myself better. In the other blockchains where I have these contracts deployed (polygon, eth, binance, avalanche). The OZ plugins for Hardhat or Truffle are used for the use of proxies.
In the architecture we use (erc721 contract factory) we have a transparent proxy for the factory, and then a beacon proxy for the token contracts.
Question, in this test that I have done with your help, I have only deployed the factory contract by generating the bytecode with solc-js directly.
Now I have tried with the bytecode generated by Truffle as Hardhat.

And I get this message:

```
ReceiptStatusError: receipt for transaction 0.0.0@1656489717.954698316 contained error status CONTRACT_REVERT_EXECUTED
```

Is there any problem with the bytecodes generated by these frameworks?

**Reply by damianlluch** [reply]
I have been able to deploy contracts with the ABI's generated by truffle - hardhat.
Using the "bytecode" field of the compile json.
Now, I have the issue of proxies.
I need to deploy a transparent proxy beacon and a transparent proxy, with the other blockchain's I use the OZ plugins that do all the migration behind the scenes.
In this case, I have to do all the proxy deployment manually.

Has anyone done something similar?
